binr/ragg2/ragg2.c

```diff
@@ -108,6 +109,257 @@ static int openfile (const char *f, int x) {
+		if (!strcmp (arch, "x86")) {
+			if (bits == 32) {
+				strcpy (CFLAGS, "-fPIC -fPIE -pie -fpic -m32");
+				strcpy (LDFLAGS, "-fPIC -fPIE -pie -fpic -m32");
+			} else {
+				strcpy (CFLAGS, "-fPIC -fPIE -pie -fpic -m64");
+				strcpy (LDFLAGS, "-fPIC -fPIE -pie -fpic -m64");
+			}
+		} else {
+			strcpy (CFLAGS, "-fPIC -fPIE -pie -fpic -nostartfiles");
+			strcpy (LDFLAGS, "-fPIC -fPIE -pie -fpic -nostartfiles");
+		}
+	}
+
+	snprintf (TRIPLET, sizeof (TRIPLET), "%s-%s-%d", os, arch, bits);
+
+	if (!strcmp (os, "windows")) {
+		strcpy (TEXT, ".text");
+		strcpy (FMT, "pe");
+	} else if (!strcmp (os, "darwin")) {
+		strcpy (TEXT, "0.__TEXT.__text");
+	} else {
+		strcpy (TEXT, ".text");
+	}
+		
+	use_clang = false;
+	if (!strcmp (TRIPLET, "darwin-arm-64")) {
+		strcpy (CC, "xcrun --sdk iphoneos gcc -arch arm64 -miphoneos-version-min=0.0");
+		use_clang = true;
+		strcpy (TEXT, "0.__TEXT.__text");
+	} else if (!strcmp (TRIPLET, "darwin-arm-32")) {
+		strcpy (CC, "xcrun --sdk iphoneos gcc -arch armv7 -miphoneos-version-min=0.0");
+		use_clang = true;
+		strcpy (TEXT, "0.__TEXT.__text");
+	}
+
+	snprintf (buffer, sizeof (buffer), "%s -nostdinc -include %s/%s/sflib.h",
+	  			CFLAGS, SFLIBPATH, TRIPLET);
+	strncpy (CFLAGS, buffer, sizeof (CFLAGS) - 1);
+
+	if (use_clang) {
+		snprintf (buffer, sizeof (buffer), "%s -fomit-frame-pointer"
+		  				" -fno-zero-initialized-in-bss", CFLAGS);
+		strncpy (CFLAGS, buffer, sizeof (CFLAGS) - 1);
+	} else { 
+		snprintf (buffer, sizeof (buffer), "%s -z execstack -fomit-frame-pointer"
+		  				" -finline-functions -fno-zero-initialized-in-bss", CFLAGS);
+		strncpy (CFLAGS, buffer, sizeof (CFLAGS) - 1);
+	}
+	snprintf (buffer, sizeof (buffer), "%s -nostdlib", LDFLAGS);
+	strncpy (LDFLAGS, buffer, sizeof (LDFLAGS) - 1);
+
+	printf ("==> Compile\n");
+	printf ("%s %s -o %s.tmp -S -Os %s\n", CC, CFLAGS, file, file);
+
+	free (output);
+	output = r_sys_cmd_strf ("(%s %s -o %s.tmp -S -Os %s) 2>&1", CC, CFLAGS, file, file);
+	if (!output) {
+		eprintf ("Compilation failed!\n");
+		goto fail;
+	}
+	printf ("%s", output);
+
+	if ((fileExt = malloc (strlen (file) + 6)) == NULL) {
+		eprint ("Malloc for fileExt failed!\n");
+		goto fail;
+	}
+
+	sprintf (fileExt, "%s.s", file);
+
+	fileExtFd = open (fileExt, O_CREAT | O_RDWR, 0644);
+	if (fileExtFd == -1) {
+		eprintf ("Error while opening %s.s\n", file);
+		goto fail;
+	}
+	write (fileExtFd, SHDR, strlen (SHDR));
+	close (fileExtFd);
+
+	free (output);
+	output = r_sys_cmd_strf ("cat %s.tmp | sed -e s,rdata,text, -e 's,get_pc_thunk.bx,__getesp__,g' "
+	  				 "| grep -v .cstring | grep -v size | grep -v ___main | grep -v section "
+					 "| grep -v __alloca | grep -v zero | grep -v cfi >> %s.s", file, file);
+	if (!output) {
+		eprintf ("Write in %s.s failed!\n", file);
+		goto fail;
+	}
+	printf ("%s", output);
+
+	printf ("==========================\n");
+	
+	printf ("==> Assemble\n");
+	printf ("%s %s -Os -o %s.o %s.s\n", CC, LDFLAGS, file, file);
+
+	free (output);
```

Comment:
Command injection and will fail with path with spaces. U may want to use the argv api to escape argumwnts

Reply:
Ok, i'm on it!